File: ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) <= 1:
            return False

        brackets = {")" : "(", "}" : "{", "]" : "["}
        stack = []

        #Make sure the opening brackets correspond with its closing brackets
        for char in s:
            if not char in brackets:
                stack.append(char)
            elif stack and brackets[char] == stack[-1]:
                stack.pop()
            else:
                return False
        
        # #Make sure all opening brackets has a closing bracket
        if len(stack) == 0:
            return True
        else:
            return False
            
        #Time: O(n)
        #Space: O(n)

File: ValidParentheses.java
class Solution {
    public boolean isValid(String s) {
        
        Map<Character, Character> brackets = new HashMap<Character, Character>();
        brackets.put(')', '(');
        brackets.put('}', '{');
        brackets.put(']', '[');

        Stack<Character> stack = new Stack<Character>();

        for (char i : s.toCharArray()){
            if (!brackets.containsKey(i)){
                stack.push(i);
            }
            else if (stack.size() > 0 && stack.peek() == brackets.get(i)){
                stack.pop();
            }else{
                return false;
            }
        
        }

        if (stack.size() != 0){
            return false;
        }
        return true;
    }

    // Time: O(n)
    // Space: O(n)
}
